File: backend/app/schemas/ticket.py
from pydantic import BaseModel, Field, EmailStr
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum

# Enums
class TicketSource(str, Enum):
    EMAIL = "email"
    SMS = "sms"
    GLPI = "glpi"
    SOLMAN = "solman"
    WEB = "web"
    PHONE = "phone"

class TicketUrgency(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

class TicketStatus(str, Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    PENDING = "pending"
    RESOLVED = "resolved"
    CLOSED = "closed"

class InteractionType(str, Enum):
    USER_MESSAGE = "user_message"
    AI_RESPONSE = "ai_response"
    AGENT_RESPONSE = "agent_response"
    SYSTEM_NOTE = "system_note"
    EMAIL_SENT = "email_sent"
    SMS_SENT = "sms_sent"

# Request Schemas
class TicketCreate(BaseModel):
    source: TicketSource = TicketSource.WEB
    user_email: str = Field(..., min_length=5, max_length=254)
    user_phone: Optional[str] = Field(None, max_length=20)
    user_name: Optional[str] = Field(None, max_length=100)
    subject: str = Field(..., min_length=5, max_length=200)
    description: str = Field(..., min_length=10, max_length=5000)
    urgency: TicketUrgency = TicketUrgency.MEDIUM
    category: Optional[str] = None
    external_id: Optional[str] = None  # For external system integration

class TicketUpdate(BaseModel):
    status: Optional[TicketStatus] = None
    assigned_to: Optional[str] = None
    assigned_group: Optional[str] = None
    resolution_notes: Optional[str] = None
    category: Optional[str] = None
    urgency: Optional[TicketUrgency] = None
    tags: Optional[str] = None

class InteractionCreate(BaseModel):
    interaction_type: InteractionType
    content: str = Field(..., min_length=1, max_length=5000)
    sender: Optional[str] = None
    recipient: Optional[str] = None
    is_internal: bool = False

# Response Schemas
class TicketInteractionResponse(BaseModel):
    id: str
    ticket_id: str
    interaction_type: str
    content: str
    sender: Optional[str]
    recipient: Optional[str]
    is_internal: bool
    created_at: datetime
    
    class Config:
        from_attributes = True

class TicketAttachmentResponse(BaseModel):
    id: str
    filename: str
    original_filename: str
    file_size: int
    content_type: str
    uploaded_by: Optional[str]
    created_at: datetime
    
    class Config:
        from_attributes = True

class TicketResponse(BaseModel):
    id: str
    ticket_number: str
    external_id: Optional[str]
    source: str
    user_email: str
    user_phone: Optional[str]
    user_name: Optional[str]
    subject: str
    description: str
    language: str
    category: Optional[str]
    subcategory: Optional[str]
    urgency: str
    priority: int
    status: str
    assigned_to: Optional[str]
    assigned_group: Optional[str]
    ai_processed: bool
    ai_response: Optional[str]
    ai_confidence: float
    resolution_notes: Optional[str]
    resolution_time_minutes: Optional[int]
    created_at: datetime
    updated_at: datetime
    resolved_at: Optional[datetime]
    closed_at: Optional[datetime]
    tags: Optional[str]
    interactions: List[TicketInteractionResponse] = []
    attachments: List[TicketAttachmentResponse] = []
    
    class Config:
        from_attributes = True

class TicketSummary(BaseModel):
    id: str
    ticket_number: str
    subject: str
    status: str
    urgency: str
    category: Optional[str]
    created_at: datetime
    user_email: str
    ai_processed: bool

# Analytics Schemas
class AnalyticsResponse(BaseModel):
    total_tickets: int
    open_tickets: int
    in_progress_tickets: int
    resolved_tickets: int
    closed_tickets: int
    average_resolution_time_hours: float
    tickets_by_category: Dict[str, int]
    tickets_by_urgency: Dict[str, int]
    tickets_by_source: Dict[str, int]
    ai_processing_rate: float
    recent_activity: List[Dict[str, Any]]

# Search Schemas
class SearchRequest(BaseModel):
    query: str = Field(..., min_length=3, max_length=500)
    limit: int = Field(5, ge=1, le=20)
    category_filter: Optional[str] = None
    status_filter: Optional[TicketStatus] = None
    urgency_filter: Optional[TicketUrgency] = None

class SimilarTicketResponse(BaseModel):
    ticket_id: str
    ticket_number: str
    subject: str
    category: Optional[str]
    similarity_score: float
    status: str
    resolution_notes: Optional[str]

File: frontend/src/types/ticket.ts
// TypeScript interfaces for the Ticket Agent system

export enum TicketSource {
  EMAIL = 'email',
  SMS = 'sms',
  GLPI = 'glpi',
  SOLMAN = 'solman',
  WEB = 'web',
  PHONE = 'phone'
}

export enum TicketUrgency {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical'
}

export enum TicketStatus {
  OPEN = 'open',
  IN_PROGRESS = 'in_progress',
  PENDING = 'pending',
  RESOLVED = 'resolved',
  CLOSED = 'closed'
}

export enum InteractionType {
  USER_MESSAGE = 'user_message',
  AI_RESPONSE = 'ai_response',
  AGENT_RESPONSE = 'agent_response',
  SYSTEM_NOTE = 'system_note',
  EMAIL_SENT = 'email_sent',
  SMS_SENT = 'sms_sent'
}

export interface TicketInteraction {
  id: string;
  ticket_id: string;
  interaction_type: InteractionType | string;
  content: string;
  sender?: string;
  recipient?: string;
  is_internal: boolean;
  created_at: string;
}

export interface TicketAttachment {
  id: string;
  filename: string;
  original_filename: string;
  file_size: number;
  content_type: string;
  uploaded_by?: string;
  created_at: string;
}

export interface Ticket {
  id: string;
  ticket_number?: string;
  external_id?: string;
  source: TicketSource | string;
  user_email?: string;
  user_phone?: string;
  user_name?: string;
  subject?: string;
  text: string; // For compatibility with existing code
  description?: string;
  language: string;
  category?: string;
  subcategory?: string;
  urgency: TicketUrgency | string;
  priority?: number;
  status: TicketStatus | string;
  assigned_to?: string;
  assigned_group?: string;
  ai_processed?: boolean;
  ai_response?: string;
  ai_confidence?: number;
  resolution_notes?: string;
  resolution_time_minutes?: number;
  created_at: string;
  updated_at: string;
  resolved_at?: string;
  closed_at?: string;
  tags?: string;
  interactions?: TicketInteraction[];
  attachments?: TicketAttachment[];
}

export interface CreateTicketRequest {
  source: TicketSource;
  user_email: string;
  user_phone?: string;
  user_name?: string;
  subject: string;
  description: string;
  urgency?: TicketUrgency;
  category?: string;
  external_id?: string;
}

export interface UpdateTicketRequest {
  status?: TicketStatus;
  assigned_to?: string;
  assigned_group?: string;
  resolution_notes?: string;
  category?: string;
  urgency?: TicketUrgency;
  tags?: string;
}

export interface DashboardStats {
  total_tickets: number;
  open_tickets: number;
  in_progress_tickets?: number;
  resolved_tickets: number;
  closed_tickets?: number;
  avg_resolution_time: number;
  success_rate: number;
  categories: Record<string, number>;
  urgency_breakdown: Record<string, number>;
  ai_processing_rate?: number;
  recent_activity?: any[];
}

export interface AnalyticsResponse extends DashboardStats {
  average_resolution_time_hours: number;
  tickets_by_category: Record<string, number>;
  tickets_by_urgency: Record<string, number>;
  tickets_by_source: Record<string, number>;
}

export interface SearchRequest {
  query: string;
  limit?: number;
  category_filter?: string;
  status_filter?: TicketStatus;
  urgency_filter?: TicketUrgency;
}

export interface SimilarTicketResponse {
  ticket_id: string;
  ticket_number: string;
  subject: string;
  category?: string;
  similarity_score: number;
  status: string;
  resolution_notes?: string;
}

export interface APIResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface SubmitTicketResponse {
  id: string;
  ticket_number: string;
  ai_response?: string;
  category?: string;
  urgency?: string;
  status: string;
  created_at: string;
}

// Form interfaces for React components
export interface TicketFormData {
  text: string;
  source: string;
  user_email: string;
  user_name?: string;
  user_phone?: string;
  urgency?: string;
  category?: string;
}

export interface MessageFormData {
  text: string;
  user_email?: string;
  user_name?: string;
}

// Component props interfaces
export interface TicketFormProps {
  onSubmit?: (data: TicketFormData) => Promise<void>;
  initialData?: Partial<TicketFormData>;
  isLoading?: boolean;
}

export interface DashboardProps {
  refreshInterval?: number;
  showRealTimeUpdates?: boolean;
}

// Error interfaces
export interface APIError {
  message: string;
  field?: string;
  details?: Record<string, any>;
}

export interface ValidationError extends APIError {
  field: string;
}

// Hook return types
export interface UseTicketsReturn {
  tickets: Ticket[];
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
  createTicket: (data: CreateTicketRequest) => Promise<Ticket>;
  updateTicket: (id: string, data: UpdateTicketRequest) => Promise<Ticket>;
}

export interface UseAnalyticsReturn {
  stats: DashboardStats | null;
  loading: boolean;
  error: string | null;
  refresh: () => Promise<void>;
}